### PR TITLE
feat: Windows integrity check

### PR DESCRIPTION
**Pull request by guohaolay**
#### Description of Change

This PR adds integrity check for Electron applications on Windows by invoking the integrity checker on the main executable. The integrity checker verifies the checksums of the files defined on the Windows executable resources. We designed this PR using a non-intrusive and flexible architecture to facilitate review and integration.

Sample usage workflow:

1. Developers add an integrity check config file as a Windows executable resource (type: Integrity, item: Checksums).
    1. Windows executable resource ref: https://learn.microsoft.com/en-us/windows/win32/menurc/resources
    2. Developers can modify existing executable resources using tools like Resource Hacker or PowerShell scripts (https://github.com/WeiyunD/WinExeResourceEditor), which don't require recompiling the executable source code.
2. Developers sign the Electron application executable with their business certificate.
3. When the Electron application starts, the Windows integrity check in this PR will:
    1. Read the integrity check config file in the executable resource.
    2. Verify the checksum for every single file listed in the integrity check config.
    3. Exit the application if the checksums don't match.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Added Windows integrity check feature into Electron

Proposal
----

#### Authors
[Weiyun Dai](https://github.com/WeiyunD) and [Guo Hao Lay](https://github.com/guohaolay) from Amazon Games.

#### Context

The Electron applications currently lacks the integrity check feature on the Windows platform. By verifying the integrity of the application's files and resources (especially the ASAR file), we can prevent unintended modifications and tampering. This is essential in protecting users from potential security threats, such as malware injection or data breaches. With integrity check, developers can ensure the applications are genuine and unaltered, as well as providing the users with confidence in the application's reliability and trust in the company brand and reputation.

#### Workflow

This PR enables the Electron applications integrity check on Windows using the following workflow:

1. Developers add an integrity check config file as a Windows executable resource (type: Integrity, item: Checksums).
    1. Windows executable resource ref: https://learn.microsoft.com/en-us/windows/win32/menurc/resources
    2. Developers can modify existing executable resources using tools like Resource Hacker or PowerShell scripts (https://github.com/WeiyunD/WinExeResourceEditor), which don't require recompiling the executable source code.
2. Developers sign the Electron application executable with their business certificate.
3. When the Electron application starts, the Windows integrity check in this PR will:
    1. Read the integrity check config file in the executable resource.
    2. Verify the checksum for every single file listed in the integrity check config.
    3. Exit the application if the checksums don't match.

#### Integrity Check Config File Format

The integrity check config file format is a plaintext file without any advanced formatting like XML or JSON. Each integrity check entry contains three lines:

1. File path (supports only files in the sub-directories relative path).
2. Hash algorithm (supports Windows CNG algorithms: https://learn.microsoft.com/en-us/windows/win32/seccng/cng-algorithm-identifiers).
3. Hash value (case-insensitive).

There is no limit on the number of integrity check entries in the config. Developers can leverage this integrity check feature to validate any application files. Here is a example:

```
dependency1.dll
SHA256
{SHA256 value of the dependency1.dll}
resource\app.asar
SHA512
{SHA512 value of the resource\app.asar}
```
#### Testing and Performance

We have tested the integrity check code with file sizes ranging up to 1GB using a 32MB chunk for file read on an Intel(R) Core(TM) i9-10980HK CPU @ 2.40GHz 3.10 GHz with 32GB ram. Here are the performance data points:

| Files | File size | Total time (ms) | Read time (ms) | Hash time (ms) |
| ----- | --------- | --------------- | -------------- | -------------- |
|     1	|      32MB	|              95 |             26 |             69 |
|     1	|      64MB	|             180 |             50 |            134 |
|     1	|     128MB |             368 |             98 |            269 |
|     1	|     256MB	|             789 |            203 |            586 |
|    10	|     256MB	|            7206 |           3561 |           3645 |
|     1	|      1GB	|            3664 |           1010 |           2654 |

Based on the performance data, we can make the following conclusions:

* File sizes up to 128MB are considered very acceptable.
* A file size of 256MB remains within acceptable limits.
* However, a 1GB file is deemed unacceptable due to the significant time cost.

Therefore, for resource files larger than 256MB, we recommend the following approach:

* Implement the integrity check for the first loaded ASAR file (or DLL) as it maintains satisfactory performance.
* For additional integrity checks on larger files, consider implementing a custom integrity check solution inside the ASAR (or DLL) to execute checks in a background thread. This approach ensures efficient integrity checks for larger resource files without affecting the user experience.

#### To align with Electron maintainers

We (both authors) are not Electron experts. We have raised this PR because we believe the integrity check is a crucial feature for our end users. We recognize this PR has many potential areas for improvement and we would like to collaborate and align on the proposed integrity check approach regarding the following open questions:

* strategy: adding resources and integrity check code to Electron executable
* allowed file paths: absolute file path on the end user PC or limit to within the application directory. Absolute file path may create weird dependency making application fragile.
* file size limitation and chunk size: static or dynamic chunk size depending on file size could improve performance when generating checksums
* invocation: best place to invoke the integrity check
* tests: the tests that are required to make this PR merge
* documentation: the documentation/tutorial that are needed to make this PR merge

Our goal is to add the integrity check feature for Windows Electron applications in a simple, non-intrusive way to get better adoption with the Electron authors, however we are totally open to collaborate on this feature to meet the teams requirements.

#### FAQs

**What are the external dependencies in the Windows integrity check code? What's the potential maintenance effort on upgrading the dependencies?**

The Windows integrity check code uses Windows CNG (Cryptography API: Next Generation) to calculate the hash value of the files. It only has dependencies on Windows native libraries (Bcrypt.dll, Crypt32.dll, User32.dll). The libraries will be updated with Windows OS updates, and zero maintenance effort is required from Electron engineers in the future. This is also the reason we don't use any advanced text format (XML or JSON) for the integrity check config file.

**Why use Windows executable resources to store the integrity check config?**

The integrity check must be done by the main executable, and the config must be part of the executable. Other approaches like implementing integrity checks in a separate DLL are not viable since attackers can tamper with the DLL as well. Saving the integrity check config outside the executable will also give attackers the capability to tamper with the config. Adding the integrity check config into Windows executable resources in our opinion is the best solution that satisfies all the following criteria:

1. Developers can update the config without recompiling the Electron executable.
2. The config is part of the executable. Tampering with the config will break the existing executable signature.
3. Resource modification can be easily built into existing build workflows (ie Forge, Electron builder...etc) and allow developers to include other files of their choice.

**What if the Electron executable is tampered with?**

We recommend all Electron application executable to be signed after resource modification. A tampered Electron executable will break the existing executable certificate and change the executable checksum, which will trigger Microsoft Defender SmartScreen (ref: https://learn.microsoft.com/en-us/windows/security/operating-system-security/virus-and-threat-protection/microsoft-defender-smartscreen/).

**If a customer's PC is already tampered with, how much will this integrity check help?**

This integrity check will raise the current security bar from Windows "everyone"/"authorized user" to "administrator". This means, attackers need to first turn off Microsoft Defender SmartScreen or inject their certificate into the PC's trusted certificate store in order to bypass Windows executable screening and our Electron application integrity check. Considering environments like internet cafes (or any other public sharing computers), anyone can modify the ASAR file and upload confidential data (like duplicating any web requests and any accessible files on the PC) to attackers' secret service. This integrity check PR will mitigate the risks as long as attackers don't have administrator privileges on the PC.

**Will the Windows integrity check code support dynamic checksums, such as pulling from a given website?**

At present, there is no plan to support dynamic checksums, such as pulling from a website/server, this strategy are already available on the market, we are not trying to replace these existing solution. The goal of this PR is to empower Electron developers to have better protection for their applications.
If the developer wishes to use dynamic checksums, they could still achieve that with this PR by first verifying their ASAR, which will then perform dynamic checksums for all other resources.

### CODE REVIEW

**Review by MarshallOfSound — COMMENTED** [review]
Hey @guohaolay, thanks for the PR and the incredible amount of detailed notes that came with it, they provide a lot of very useful context.

Before we dive too deep into actual code review I'm wondering if y'all considered adopting support for ASAR Integrity on Windows in a similar way to macOS.

* [ASAR Integrity Docs](https://www.electronjs.org/docs/latest/tutorial/asar-integrity)
  * The integrity hashes for ASAR files are loaded on macOS from the [Info.plist](https://github.com/electron/electron/blob/main/shell/common/asar/archive_mac.mm#L35-L65) which is a code signature validated file (similar to your resources trick on windows)
  * It's currently only supported on macOS for trust chain reasons, at the time (and kind of still) there wasn't a massive benefit to similar tech on Windows/Linux as the OS isn't that great at validating the actual apps code signature.  That said your reasoning in this PR is perfectly valid for enterprise / controlled machine usescases (e.g. validating apps via applocker)

My suggestion here would be to instead of implementing all of this again separately for windows to instead just provide an implementation of `Archive::HeaderIntegrity` for windows (currently there isn't one) and then the platforms are perfectly aligned.

ASAR Integrity also solves the perf-penalty issues you noted somewhat as it only validates specific chunks on demand and does threaded stream based validation for files loaded into renderers from disk (E.g. think large video files)

If y'all want to chat about this in a more direct manner also please feel free to reach out to my email address (my github username @ electronjs{dot}org) and we can look at getting y'all into our Slack or something

**Comment by WeiyunD** [comment]
Hi @MarshallOfSound ,

It's Weiyun, co-author of this PR. Thanks for the quick response. It's great to see that the Electron maintainer group is open to considering this feature.

Your breakdown makes a lot of sense. Essentially, we're aiming for the integrity check to behave the same way on both Windows and Mac, with the only difference being where the checksum value is loaded from.

We'll dive into the existing Mac version of `Archive::HeaderIntegrity` and follow up with you directly afterward.

Cheers, Weiyun

**Comment by MarshallOfSound** [comment]
@zcbenz We're discussing this atm in #proj-asar-integrity-windows

File: shell/common/asar/archive_win.cc
Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:29]: Blocking: Shouldn't this be `FATAL`? If the fuse is enabled and this part of the logic fails we should error on the side of caution and prevent the file read, so somehow this needs to fail, not silent pass.

File: shell/common/asar/archive_win.cc
Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:39]: Blocking: Same here

File: shell/common/asar/archive_win.cc
Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:54]: Blocking: Instead of using an empty string can we stick we `absl::optional`

Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:54]: Perf Nit: To avoid constant lookups should this be a static `base::NoDestructor`? Otherwise we recalculating this for every ASAR that is opened

File: shell/common/asar/archive_win.cc
Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:93]: I don't think we need `__func__` in these logs, Chrome doesn't do it and the LOG macro already has line numbers and filenames

File: shell/common/asar/archive_win.cc
Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:85]: Question: Is there a reason the RCData is parsed out with this very specific multiline format, instead of just using JSON and a standard parser? Would be way easier to read and understand as an implementer if we used a standard data format

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc:153]: Changed to JSON. Sample

```
[
  {
    "file": "resources\\data.asar",
    "alg": "1",
    "value": "22222"
  },
  {
    "file": "resources\\default_app.ASAR",
    "alg": "sha256",
    "value": "dd29126b7b8a5959bc93236da70d709333a9c7522d5b9cda5b0e72b73ca9943b"
  }
]
```

File: shell/common/asar/archive_win.cc
Comment (MarshallOfSound) [inline shell/common/asar/archive_win.cc:149]: I think this should be moved?

```suggestion
        std::move(header_integrity);
```

**Review by MarshallOfSound — APPROVED** [review]
One nit RE a `move` that I'm not sure about, otherwise LGTM

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:81]: Chromium has `PLOG` for automatically printing out the error:
```suggestion
    PLOG(FATAL) << "FindResource failed";
```

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:95]: `res_size` being 0 is a distinct error, perhaps it should get a different log message? Also, we should probably `PLOG` if `LockResource` returns nullptr.

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:92]: It's probably fine to pass `NULL` here as `module_handle` is the current exe's module which is the default here, but it can't hurt to pass `module_handle` in directly since we have it right here :)

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:20]: Avoid `using namespace` in Electron code.

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:100]: nit:
```suggestion
  std::string integrity_config_payload(res_data, res_size);
```

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:103]: Since the JSON in the integrity blob is machine-generated and not intended to be read by the user, is there any reason to allow comments/commas/etc.?

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc:149]: Will test it

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:166]: Are we guaranteed that the relative path is ASCII? `base::WideToUTF8` might be more appropriate here. If it really is important that it be ASCII, this fact deserves a comment.

Also, `.c_str()` isn't required as `base::WideTo*` works on `std::wstring_view`.
```suggestion
  string rel_path_utf8 = base::WideToUTF8(relative_path.value().value());
```

**Review by nornagon — CHANGES_REQUESTED** [review]

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated and tested

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated. Thank you for guiding me to the correct function. I'm impressed by the comprehensiveness of your utility function.

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated. Split the if statement to two and used `PLOG` for both

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Good catch. Updated.

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated.

I understand why we should avoid using namespace in header files, but curious to know why also avoid doing so in source files.

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Updated

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: You are correct. I should convert to UTF8. Updated.

File: shell/common/asar/archive_win.cc
Comment (nornagon) [reply shell/common/asar/archive_win.cc]: These are from Chromium. Chromium does a lot of stuff 😅

File: shell/common/asar/archive_win.cc
Comment (nornagon) [reply shell/common/asar/archive_win.cc]: The main reason is that it makes it harder to know where a symbol comes from when reading the file. It's also against the Chromium/Google C++ style guide to do so: https://google.github.io/styleguide/cppguide.html#Namespaces

File: shell/common/asar/archive_win.cc
Comment (nornagon) [inline shell/common/asar/archive_win.cc:45]:
```suggestion
      integrity_config_cache;
```

**Review by nornagon — APPROVED** [review]
lgtm after spelling fix

File: shell/common/asar/archive_win.cc
Comment (WeiyunD) [reply shell/common/asar/archive_win.cc]: Good catch. Will fix it.

**Comment by WeiyunD** [comment]
Hi @MarshallOfSound @nornagon

Thank you once more for your support and review of this PR. It appears that the current changes have successfully passed all the required approval steps. I believe the Linux build failure is unrelated to my changes.

Is there anything more I need to do on my end to proceed with merging the PR?

**Comment by WeiyunD** [comment]
The PR has undergone several updates based on feedback from Electron maintainers. Below are the key updates:

1. **Checksum Algorithm**:
   - Electron currently only supports SHA256 for Mac, hence we are limiting to SHA256 support as well.

2. **Checksum Content**:
   - To optimize performance, especially for large ASAR files, we have modified our approach. Instead of hashing the entire ASAR file, we now only hash the raw ASAR header, aligning with Electron’s existing practice on Mac.
   - Here is a JavaScript example demonstrating how to obtain the checksum of an ASAR header:
     ```javascript
     const asar = require('@electron/asar');
     const crypto = require("crypto");

     function getChecksum(str) {
         return crypto.createHash("sha256").update(str).digest("hex");
     }

     const header = asar.getRawHeader('default_app.asar');
     console.log(getChecksum(header.headerString));
     ```

3. **Windows Integrity Config Format**:
   - The format has been updated to JSON. Below is a sample of the new format:
     ```json
     [
       {
         "file": "resources\\data.asar",
         "alg": "sha256",
         "value": "..."
       },
       {
         "file": "resources\\default_app.asar",
         "alg": "sha256",
         "value": "..."
       },
       /* ... more asar files ... */
     ]
     ```

4. **Windows Integrity Config Resource Type & Item**:
   - The JSON configuration for integrity checks must be stored in the Windows executable resource with:
     - Resource Type: `Integrity`
     - Resource Item: `ElectronAsar`

**Comment by guohaolay** [comment]
> @guohaolay Maybe it is a good idea to allow encrypted ASAR files as well that are loaded with the same encryption key as the integrity validation and fail if none is present. This to prevent hackers from replacing the whole electron executable with one without any integrity checks embedded as resource like the original one from the electron distribution.

@gpetrov that's a great call out, currently I don't believe electron natively generates encrypted ASAR hence this is not our main focus.

Regarding protection against the replacement of the electron executable, when this happens, it is likely something more devastating has compromised the computer security and the hacker has higher privilege and admin access, therefore at this point there isn't really anything we can do to protect the executable.

**Comment by lrs955** [comment]
Is there some plans to extend the integrity check also on Linux?

**Pull request merged**
